syntax = "proto3";

package proxy;

import "google/protobuf/empty.proto";
import "dspike/proto/common.proto";

// Proxy Interface for spike messages
service Proxy {
  // do TCP transfer for Core <-> LLB
  rpc TCPXfer(dspike.proto.Message) returns (google.protobuf.Empty);
  // do bsp sync
  rpc Sync(SyncRequest) returns (google.protobuf.Empty);

  // do DMA transfer for LLB <-> DDR
  rpc DMAXfer(dspike.proto.Message) returns (google.protobuf.Empty);
  // poll DMA transfer status
  rpc DMAXferPoll(google.protobuf.Empty) returns (dspike.proto.DMAXferPollResponse);

  // receive messages via stream
  rpc RecvCb(stream RecvCbRequest) returns (stream dspike.proto.Message);

  // upload dump data to proxy
  rpc Dump(DumpData) returns (google.protobuf.Empty);
  // wait for proxy send a DumpParam to get dump data or turn on/off sync dump
  rpc WaitDump(WaitDumpRequest) returns (stream dspike.proto.DumpParam);
}

// RecvCbRequest present a request for spike to receive messages
message RecvCbRequest {
  int32 spikeId = 1;
}

// SyncRequest present a request for spike to do sync operation
message SyncRequest {
  int32 spikeId = 1;
}

// DumpData present a dump data of spike
message DumpData {
  dspike.proto.DumpInfo info = 1;
  bytes data = 2;
}

// WaitDumpRequest present a request for spike to receive dump param
message WaitDumpRequest {
  int32 spikeId = 1;
}